src/bosh-director/spec/unit/problem_resolver_spec.rb

```diff
@@ -31,96 +50,201 @@ def inactive_disk(id, deployment_id = nil)
+              expect(ThreadPool).to have_received(:new).once.with(max_threads: n_igs_with_problems)
+              expect(ThreadPool).to have_received(:new).exactly(n_igs_with_problems).times.with(max_threads: n_problems_in_ig)
+            end
+          end
+
+          context 'when max_in_flight is smaller than the number of instances with problems and max_threads' do
+            let(:max_in_flight) { 2 }
 
-          2.times do
-            disk = Models::PersistentDisk.make(active: false)
-            disks << disk
-            problems << inactive_disk(disk.id)
+            it 'respects max_in_flight' do
+              test_instance_apply_resolutions
+              expect(ThreadPool).to have_received(:new).once.with(max_threads: n_igs_with_problems)
+              expect(ThreadPool).to have_received(:new).exactly(n_igs_with_problems).times.with(max_threads: max_in_flight)
+            end
           end
 
-          resolver = make_resolver(@deployment)
+          context 'when max_threads is smaller than the number of instances with problems and max_in_flight' do
+            let(:max_threads) { 2 }
+            it 'respects max_threads' do
+              test_instance_apply_resolutions
+              expect(ThreadPool).to have_received(:new).exactly(1 + n_igs_with_problems).times.with(max_threads: max_threads)
+            end
+          end
 
-          expect(resolver).to receive(:track_and_log).with(/Disk 'disk-cid-\d+' \(0M\) for instance 'job-\d+\/uuid-\d+ \(\d+\)' is inactive \(.*\): .*/).twice.and_call_original
+          context 'when serial is true for some instance groups' do
+            let(:serial_update_config) { instance_double('Bosh::Director::DeploymentPlan::UpdateConfig', serial?: true) }
+            let(:igs) do
+              [
+                instance_double('Bosh::Director::DeploymentPlan::InstanceGroup', name: 'ig-1', update: serial_update_config),
+                instance_double('Bosh::Director::DeploymentPlan::InstanceGroup', name: 'ig-2', update: parallel_update_config),
+                instance_double('Bosh::Director::DeploymentPlan::InstanceGroup', name: 'ig-3', update: parallel_update_config),
+                instance_double('Bosh::Director::DeploymentPlan::InstanceGroup', name: 'ig-4', update: serial_update_config),
+              ]
+            end
+            before do
+              allow(serial_update_config).to receive(:max_in_flight).and_return(max_in_flight)
+            end
 
-          expect(resolver.apply_resolutions(problems[0].id.to_s => 'delete_disk', problems[1].id.to_s => 'ignore'))
-            .to eq([2, nil])
+            it 'respects serial' do
+              test_instance_apply_resolutions
+              non_serial_igs_with_problems = 2
+              expect(ThreadPool).to have_received(:new).once.with(max_threads: non_serial_igs_with_problems)
+              expect(ThreadPool).to have_received(:new).exactly(
+                n_igs_with_problems,
```

Comment:
`n_igs_with_problems` is currently set to 1 in this context (line 89) but my understanding is that `ig-2` and `ig-3` would both need a ThreadPool, which means that shouldn't this be called twice? Do ThreadPools get reused across instance groups with parallel updates?

Reply:
Hi @belinda-liu ,

ThreadPools are getting reused across instance groups which should get updated in parallel.

Therefore, `ig-2` and `ig-3` will share one ThreadPool because they are configured with `serial = false`. This ThreadPool is created once and will use 2 threads to handle those instance groups in parallel.
This is what the first `expect` is testing. It checks if a ThreadPool is created once with `max_thread: 2`.

The second `expect` is testing the parallel resolution of problems inside a single instance group. Each instance group (4 in total) has 3 problems which should all get resolved in parallel. Therefore, we expect to create 4 ThreadPools with each having `max_thread: 3`.

`n_igs_with_problems` should be the default (4) again because line 89 is in another context.